pkg/routeagent_driver/handlers/ovn/submarinerNonGWRouteHandler.go

```diff
@@ -0,0 +1,199 @@
+	}
+
+	annotations := node.GetAnnotations()
+
+	// TODO transitSwitchIP changes support needs to be added.
+	transitSwitchIP, ok := annotations[constants.OvnTransitSwitchIPAnnotation]
+	if !ok {
+		logger.Infof("No transit switch IP configured")
+		return nil
+	}
+
+	h.transitSwitchIP, err = jsonToIP(transitSwitchIP)
+
+	return errors.Wrapf(err, "error parsing the transit switch IP")
+}
+
+func (h *NonGatewayRouteHandler) GetName() string {
+	return "submariner-nongw-route-handler"
+}
+
+func (h *NonGatewayRouteHandler) GetNetworkPlugins() []string {
+	// TODO enable when we switch to new implementation
+	// return []string{cni.OVNKubernetes}
+	return []string{}
+}
+
+func (h *NonGatewayRouteHandler) RemoteEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	h.remoteEndpoints[endpoint.Name] = endpoint
+
+	if !h.isGateway || h.transitSwitchIP == "" {
+		return nil
+	}
+
+	_, err := h.smClient.SubmarinerV1().
+		NonGatewayRoutes(endpoint.Namespace).Create(context.TODO(),
+		h.newNonGatewayRoute(endpoint), metav1.CreateOptions{})
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return errors.Wrapf(err, "error processing the remote endpoint create event for %q", endpoint.Name)
+	}
+
+	return nil
+}
+
+func (h *NonGatewayRouteHandler) RemoteEndpointRemoved(endpoint *submarinerv1.Endpoint) error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	delete(h.remoteEndpoints, endpoint.Name)
+
+	if !h.isGateway || h.transitSwitchIP == "" {
+		return nil
+	}
+
+	if err := h.smClient.SubmarinerV1().NonGatewayRoutes(endpoint.Namespace).Delete(context.TODO(),
+		endpoint.Name, metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrapf(err, "error deleting nonGatewayRoute %q", endpoint.Name)
+	}
+
+	return nil
+}
+
+func (h *NonGatewayRouteHandler) TransitionToNonGateway() error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	h.isGateway = false
+
+	return nil
+}
+
+func (h *NonGatewayRouteHandler) TransitionToGateway() error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	h.isGateway = true
+
+	if h.transitSwitchIP == "" {
+		return nil
+	}
+
+	for _, endpoint := range h.remoteEndpoints {
+		ngwr := h.newNonGatewayRoute(endpoint)
+
+		result, err := util.CreateOrUpdate(context.TODO(), h.nonGatewayResourceInterface(endpoint.Namespace), ngwr,
+			func(existing runtime.Object) (runtime.Object, error) {
+				existingNgwr := ngwr
+				return existingNgwr, nil
+			})
+		if err != nil {
+			return errors.Wrapf(err, "error creating/updating NonGatewayRoute")
+		}
+
+		if result == util.OperationResultCreated {
+			logger.V(log.TRACE).Infof("NonGatewayRoute does not exist - created: %+v", ngwr)
+		} else if result == util.OperationResultUpdated {
+			logger.V(log.TRACE).Infof("NonGatewayRoute already exists - updated %+v", ngwr)
+		} else {
+			logger.V(log.TRACE).Info("NonGatewayRoute already exists but doesn't need updating")
+		}
```

Comment:
Same comments apply as in the other handler.

Reply:
Done